Native + Lucene Fusion index as default

burqen opened the pull request:
## Fusion index as default
Fusion index = a fusion between native and Lucene index.
Native index is built on top of GB+Tree and can only handle numbers.
Lucene index is old Lucene implementation.

Fusion index works by wrapping Native and Lucene into one logical index where all numbers are directed to Native and the rest (including all arrays and all composite keys) to Lucene.

First benchmarks show a write performance improvement of between 1.5-4x.

## Default behaviour
- All newly created indexes will be 'fusion indexes'.
- Old indexes (created with only Lucene) will not be migrated but continue as 'Lucene only' indexes. Drop and recreate to use fusion index. This is recommended for indexes on number focused properties (not composite or arrays). For indexes over Strings there is no use to move to fusion.

## Migration
Migrating from <3.3 will keep all indexes around and will work just fine as Lucene only. Drop and recreate to move to fusion index.

## Internal config to disable
An internal unsupported config option is introduced to disable fusion index as default. In this case, Lucene will be used instead. Disabling fusion index will NOT migrate already created fusion indexes. Those will have to be dropped and recreated to use the new default (Lucene).

## TL;DR
- New fusion index = native + lucene.
- Fusion is default.
- When creating new index, default is used.
- No migration is done for already existing indexes. Can be migrated to default by drop and recreate.
- Turn fusion index of with internal unsupported config option.
- Faster, Harder, Scooter.

co-author: @tinwelint 

[cl] Introduce new native index for numbers.

tinwelint committed:
Support multiple schema index providers

Introduce kernel extension for Native+Lucene schema index provider

tinwelint committed:
Small further fixes

tinwelint committed:
Adds logging argument to native index

tinwelint committed:
Correct implementation of FusionSchemaIndexProvider#getInitialState

Commit:
More fusion index provider adaptation

- Index populator always apply updates in batches.
- Native populator use WorkSync to apply batches of updates
  single threaded.
- ConsistencyCheckService now use fusion between lucene and
  native index.

tinwelint committed:
hasFullNumberPrecision takes number arrays into consideration

tinwelint committed:
Correct implementation of getPopulationFailure in FusionSchemaIndexProvider

tinwelint committed:
Test closes updater

Commit:
Better test for FusionSchemaIndexProvider getPopulationFailure

Commit:
Assert populator updater is open and improved populator testing

Commit:
Store migration use fusion index provider

Commit:
Use fusion index provider for consistency check test

tinwelint committed:
Loads SchemaIndexProviders in consistency checker using KernelExtensions

Commit:
Introduce internal config to disable native index

Commit:
Consistency check test for old lucene schema index

tinwelint committed:
Asserts unique schema index provider descriptors

This was a problem in some tests which loaded an additional schema index provider
which happened to have the same descriptor as an existing one (loaded via extension).

Commit:
FusionSchemaIndexProviderCompatibilitySuiteTest

Native schema index handle null values as range boundary as unbounded.
Native schema index handle range with reversed range boundary as empty range.

IndexProviderCompatibilityTestSuite and friends now close opened resources.

Commit:
Test fusion index for missing files

Commit:
Dont load mulitple InMemoryIndexProviders in test

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    public static final RecoveryCleanupWorkCollector RECOVERY_PREVENTING_COLLECTOR = new RecoveryPreventingCollector();
```
Please fields first, public member and them private member

MishaDemianenko commented on community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    public static final RecoveryCleanupWorkCollector RECOVERY_PREVENTING_COLLECTOR = new RecoveryPreventingCollector();
```
not sure what collectors class and definition does in SchemaIndexExtensionLoader, even if it's passed as one of the parameters to instantiate extensions

tinwelint replied:
Well, this class is a utility used in a bunch of places in consistency checker. It loads schema index provider kernel extensions and consistency checker is not allowed to do recovery. Where should it go?

MishaDemianenko replied:
well if you will need in any other test ever RecoveryPreventingCollector I would say that SchemaIndexExtensionLoader would be the last place where i would expect to find it, even in separate class is better imho

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java:
```diff
@@ -0,0 +1,60 @@
+public class AllByPrioritySelectionStrategy<T> implements DependencyResolver.SelectionStrategy
+{
+    @SuppressWarnings( "rawtypes" )
+    private List lowerPrioritizedCandidates = Collections.emptyList();
```
`private List<T> lowerPrioritizedCandidates = Collections.emptyList();` then you do not need to suppress warnings

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java:
```diff
@@ -0,0 +1,60 @@
+ * Selects the candidate with highest priority (assumed to implement {@link Comparable}) and returns
+ * in {@link #select(Class, Iterable)}, but keeps the others for access too.
+ */
+public class AllByPrioritySelectionStrategy<T> implements DependencyResolver.SelectionStrategy
```
`<T extends Comparable>` instead `<T>`

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java:
```diff
@@ -0,0 +1,60 @@
+        return highest;
+    }
+
+    @SuppressWarnings( "unchecked" )
```
no need to suppress warnings

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public <R> R select( Class<R> type, Iterable<R> candidates ) throws IllegalArgumentException
+    {
+        List<Comparable> all = (List<Comparable>) Iterables.asList( candidates );
```
` List<T> all = (List<T>) Iterables.asList( candidates );` instead

MishaDemianenko commented on community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:
```diff
@@ -556,11 +556,13 @@ public void shouldReportNodesThatAreNotIndexed() throws Exception
             IndexUpdater updater = accessor.newUpdater( IndexUpdateMode.ONLINE );
             updater.remove( asPrimitiveLongSet( indexedNodes ) );
             updater.close();
+            accessor.force();
```
what's the point doing force specifically before closing accessor?

tinwelint replied:
That's the general contract and something that native indexing hooks in on to checkpoint properly. It's just that lucene happened to make do without it.

MishaDemianenko replied:
I mean if you close it without force it should do that automatically right?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java:
```diff
@@ -387,9 +387,11 @@ private int search( PageCursor cursor, KEY key, KEY readKey, int keyCount )
-                    cursor.getCurrentPageId();
+        long currentPageId = cursor.getCurrentPageId();
+        long expectedPageId = levels[currentLevel].treeNodeId;
+        assert currentPageId == expectedPageId : "Expected cursor to be at page:" +
```
noo, that was one of my favourite non-readable statements 😞

tinwelint replied:
Bugger, well I'd guess it wasn't the last of its kind :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -394,6 +394,10 @@
             setting( "unsupported.dbms.multi_threaded_schema_index_population_enabled", BOOLEAN, TRUE );
 
+    @Internal
+    public static final Setting<Boolean> enable_native_schema_index =
```
surprising option

tinwelint replied:
We went for it since it's easier to manage, also in tests. Now that we've gotten all fixed and working we can investigate if it would be tricky to manage this if it were a toggle instead. Hang on.

Commit:
Compilation fix after rebase

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java:
```diff
@@ -181,23 +160,10 @@ private void awaitCompletion()
      */
     private void batchUpdate( IndexPopulation population, IndexEntryUpdate<?> update )
```
this method is not used at all now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java:
```diff
@@ -206,18 +172,7 @@ private void flushIfNeeded( IndexPopulation population, List<IndexEntryUpdate<?>
-                flush( population, updates );
-            }
-        }
+        populations.forEach( population -> flush( population, population.takeCurrentBatch() ) );
```
there is no need to pass second param here since population is already passed as first one

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java:
```diff
@@ -86,6 +89,7 @@
 
     public static final String QUEUE_THRESHOLD_NAME = "queue_threshold";
     private final int QUEUE_THRESHOLD = FeatureToggles.getInteger( getClass(), QUEUE_THRESHOLD_NAME, 20_000 );
+    private final int BATCH_SIZE = FeatureToggles.getInteger( BatchingMultipleIndexPopulator.class, BATCH_SIZE_NAME, 10_000 );
```
feature toggle is moved but class is the same? why?

tinwelint replied:
Something of a confusion move at a time we were confused about multi index populator. Forgot about it... we can move it back.

MishaDemianenko replied:
The option still present in this class

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:
```diff
@@ -80,29 +78,46 @@ public IndexAccessor getOnlineAccessor( long indexId, IndexDescriptor descriptor
-        String luceneFailure = luceneProvider.getPopulationFailure( indexId );
-        if ( nativeFailure != null )
+        String nativeFailure = null;
+        try
```
what is the sense in throwing and catching useless exceptions instead of updating providers to not throw exceptions if not failed?

tinwelint replied:
`getPopulationFailure` is designed to throw `IllegalStateException` if called for index that isn't in `FAILED` state, sensible methinks.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:
```diff
@@ -80,29 +78,46 @@ public IndexAccessor getOnlineAccessor( long indexId, IndexDescriptor descriptor
+        {   // Just catch
         }
+
+        if ( nativeFailure != null || luceneFailure != null )
```
please use `StringUtils.isNotEmpty` instead

tinwelint replied:
I think we want to check `null` explicitly. Empty string means something else in this case.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:
```diff
@@ -408,9 +410,16 @@ public void start() throws IOException
+        SchemaIndexProvider defaultIndexProvider =
+                dependencyResolver.resolveDependency( SchemaIndexProvider.class, indexProviderSelection );
+
+        // todo do we really need to have this dependency?
```
TODO

tinwelint replied:
Ah, to be investigated...

tinwelint replied:
It was left there because that's how it was previously. Now the `SchemaIndexProviderMap` is used instead. It's not strictly required for us to remove this right now, since it may require some time and testing. We've created a card for it now.

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java:
```diff
@@ -1,199 +0,0 @@
-/*
```
Not quite sure why this test was removed, we do not cancel sampling anymore?

tinwelint replied:
It only tested that the in-memory test index `HashBasedIndex` sample method threw exception in a particular case. I think it was mis-written to begin with.

MishaDemianenko reviewed (changes requested):
Couple of small comments here and there

tinwelint committed:
Small touch-ups based on comments

Commit:
Made BATCH_SIZE non static to make it reloadable within same JVM

... by flipping feature toggle flag. Used in tests.
FeatureToggle flag keeps same class name to not change existing API.

tinwelint commented:
I think we need to look at `SchemaIndexHaIT` dude, it fails consisently on this branch... not very often at all otherwise. I think it's simply a kernel extension loading thingie.

Commit:
Fix SchemaIndexHaIT to not add multiple index providers

MishaDemianenko commented:
@tinwelint @burqen there is one small thing still please fix that otherwise its good and approved

tinwelint commented:
The build passed, although the Github <--> CI integration didn't quite get that bit. @burqen was there some issue with leaving open files behind (as shown in some benchmark)? If so perhaps we should hold off merge until we know more.

burqen commented:
@tinwelint It's a bit unclear. There does not seem to be any leaking files but we need some more investigation before merging.
